website/profile/views.py

```diff
@@ -786,31 +786,36 @@ def request_deactivation(auth):
     return {'message': 'Sent account deactivation request'}
 
 
-def redirect_to_twitter(uid):
-    """Redirect GET requests for /@uid/ to respective the Twitter account if a User
-    exists and has a Twitter handle associated with the account.
+def redirect_to_twitter(twitter_handle):
+    """Redirect GET requests for /@TwitterHandle/ to respective the OSF user
+    account if it associated with an active account
 
     :param uid: uid for requested User
     :return: Redirect to User's Twitter account page
     """
-    user = User.load(uid)
-
-    if user:
-        twitter_handle = user.social.get('twitter', None)
-    else:
+    try:
+        user = User.find_one(Q('social.twitter', 'iexact', twitter_handle))
+    except NoResultsFound:
         raise HTTPError(http.NOT_FOUND, data={
             'message_short': 'User Not Found',
-            'message_long': 'There is no active user associated with user id: {0}.'.format(uid)
+            'message_long': 'There is no active user associated with the Twitter handle: {0}.'.format(twitter_handle)
         })
-
-    if twitter_handle:
-        # TODO(hrybacki): Verify Twitter handle is for a real account.
-        # Requires authenticated access to Twitter API
-        redirect_url = "https://twitter.com/{0}".format(twitter_handle)
-    else:
-        raise HTTPError(http.BAD_REQUEST, data={
-            'message_short': 'Invalid Request',
-            'message_long': '{0} does not have a Twitter handle associated with their account.'.format(user.fullname)
+    except MultipleResultsFound:
+        users = User.find(Q('social.twitter', 'iexact', twitter_handle))
+        message_long = 'There are multiple OSF accounts associated with the ' \
+                       'Twitter handle: <strong>{0}</strong>. <br /> Please ' \
+                       'select from the accounts below. <br /><ul>'.format(twitter_handle)
+        for user in users:
+            message_long += '<li><a href="{0}">{1}</a></li>'.format(web_url_for(
+                'profile_view_id',
+                uid=user._id),
+                user.fullname
+            )
+        message_long += '</ul>'
+        raise HTTPError(http.MULTIPLE_CHOICES, data={
```

Comment:
I'm confused -- in this case there are multiple places the client could be
redirected to. Is this only used when multiple options exist but the server
picks the option for the client?

On Mon, Jul 6, 2015 at 2:54 PM, Steven Loria <notifications@github.com>
wrote:

> In website/profile/views.py
> <https://github.com/CenterForOpenScience/osf.io/pull/3453#discussion_r33969239>
> :
>
> > -        raise HTTPError(http.BAD_REQUEST, data={
> > -            'message_short': 'Invalid Request',
> > -            'message_long': '{0} does not have a Twitter handle associated with their account.'.format(user.fullname)
> > +    except MultipleResultsFound:
> > +        users = User.find(Q('social.twitter', 'iexact', twitter_handle))
> > +        message_long = 'There are multiple OSF accounts associated with the ' \
> > +                       'Twitter handle: <strong>{0}</strong>. <br /> Please ' \
> > +                       'select from the accounts below. <br /><ul>'.format(twitter_handle)
> > +        for user in users:
> > +            message_long += '<li><a href="{0}">{1}</a></li>'.format(web_url_for(
> > +                'profile_view_id',
> > +                uid=user._id),
> > +                user.fullname
> > +            )
> > +        message_long += '</ul>'
> > +        raise HTTPError(http.MULTIPLE_CHOICES, data={
>
> MULTIPLE_CHOICES (300) is not an appropriate response code. 300 is used
> when there are multiple places the client may be redirected to.
>
> 300 Multiple choices: There are multiple options that the client may
> follow.
>
> A 400 is more appropriate here.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/CenterForOpenScience/osf.io/pull/3453/files#r33969239>
> .
>



-- 
*Harry Rybacki* - Developer, Center for Open Science
Research: http://scinet.osf.io
Blog: http://rybacki.io
[image: http://github.com/harryrybacki]
<http://github.com/HarryRybacki> [image:
https://twitter.com/HarryRybacki] <https://twitter.com/HarryRybacki> [image:
http://www.linkedin.com/in/HarryRybacki/]
<http://www.linkedin.com/in/HarryRybacki/> [image:
https://plus.google.com/+HarryRybacki]
<https://plus.google.com/+HarryRybacki>